commit d01f6fc57ef93ae20d6dbfc44f0be565fcb808a7
Author: Josh Wilsdon <jwilsdon@joyent.com>
Date:   2019-04-11T11:03:55-07:00 (6 months ago)
    
    OS-7735 vmadmd should not break and go into maintenance when a VM exists that had a delete failure

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index 683fa2fe..30c25b74 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -16860,6 +16860,15 @@ exports.update = function (uuid, payload, options, callback)
                     }
                 }
 
+                if (disks.length === 0) {
+                    log.info({
+                        brand: vmobj.brand,
+                        uuid: uuid
+                    }, 'no disks to create, skipping createVolumes');
+                    cb();
+                    return;
+                }
+
                 // add the bits of payload createVolumes() needs.
                 p = {
                     add_disks: disks,
diff --git a/src/vm/sbin/vmadmd.js b/src/vm/sbin/vmadmd.js
index 4a8b3213..4d293465 100755
--- a/src/vm/sbin/vmadmd.js
+++ b/src/vm/sbin/vmadmd.js
@@ -2606,6 +2606,14 @@ function main()
                                 }
                             }
 
+                            if (upgrade_payload.update_nics.length === 0) {
+                                log.debug({
+                                    vm_uuid: vmobj.uuid
+                                }, 'no nics to update, skipping VM.update()');
+                                finishUpgrade(vmobj);
+                                return;
+                            }
+
                             log.info('updating ' + vmobj.uuid + ' with: '
                                 + JSON.stringify(upgrade_payload));
 
